### ISSUE TITLE
Passwords are stored in plain-text in the user's session

### ISSUE THREAD

**Issue by blubber** [Bouke/django-two-factor-auth#239]
When using the login wizard the user's password is stored in plain text in the session store. The information is stored by formtools' wizard class. This is a security risk, depending on the deployment this can easily lead to a leak. Imagine a shared database ...

The information is not required to complete the token step, it would suffice to store the authentication state once the user's username and password are verified and then call `login()` in the `done()` method.

**Comment by moggers87** [comment]
This was brought up in PR #193 (though the "fix" there just made the problem worse IMO)

A few questions:

* How long should the auth state be kept around for?
* What happens if the user's password or account status gets changed between the first and second step?

**Comment by moggers87** [comment]
This is possibly a duplicate of #221 (a different focus, but solving one would solve the other)

**Comment by blubber** [comment]
> How long should the auth state be kept around for?

I don't think it really matters that much. Around 5 minutes should be more then enough. Just s tore the user's ID, for instance in a cache entry with ttl, keyed with the session id.

> What happens if the user's password or account status gets changed between the first and second step?

You can check their status after validating the authentication token and reject the login if the account was deactivated. As long as you defer the call to `login()` until after the token has been verified I think you're ok.

**Comment by dan-klasson** [comment]
Why can't the two step authentication just be on the first form instead? That way no need to store anything in a session or continue to another page using a wizard.

**Comment by moggers87** [comment]
@dan-klasson that would work fine if we were only concerned with things like TOTP, but how would you deal with the phone call/SMS backends? We need to know who the user trying to authenticate as before we can send their device a token.

**Comment by anlutro** [comment]
An alternate approach I've seen there is to have a 2FA-code model with a foreign key to the user. That way, on the 2FA screen we can just call `auth.login(verified_2fa_code.user)` instead of `auth.authenticate`.

**Comment by blubber** [comment]
@dan-klasson you'd have to present the 2FA view to all users, regardless of the presence of a 2FA device. From a usability standpoint that would not be ideal. If your system doesn't allow users without 2FA than that would be a good solution I think.

@anlutro you'd have to check all 2FA devices in the database when a code is entered to find the right one.

The problem is not storing user information client-side during the authentication process. The problem is storing the **password** client-side. I have implemented the authentication such that the first step will verify the username/password combo, without logging in the user. The result of this authentication is stored in the cache for a couple minutes keyed with a random string. The string is also stored in a cookie on the client. Upon entering the verification code this cookie is used to retrieve authentication information from the cache and log the user in. It is somewhat cumbersome, but it is a general solution, although it does depend on the presence of a cache.

**Comment by anlutro** [comment]
> you'd have to check all 2FA devices in the database when a code is entered to find the right one.

Is that a big deal? If you index it the lookup should be easy. You could also do some GC every now and then to remove expired codes. Also, considering the 2FA token would be a separate model, you certianly wouldn't need to look up all devices to find it.

**Comment by blubber** [comment]
@anlutro you can't index that, because (at least for some devices like TOTP) the challenge is generated at the moment of verification. In the case of TOTP it includes the current time. So you'd have to generate a fresh challenge for **all** devices in the database for every login.

Also, generating a challenge for PhoneDevice will trigger an SMS or voice call.

Also see: https://bitbucket.org/psagers/django-otp/src/298ac18b69dfabf944e3ccef3cede6c91855b6a3/django-otp/django_otp/plugins/otp_totp/models.py?at=default&fileviewer=file-view-default#models.py-91

**Comment by anlutro** [comment]
My bad, we actually have an intermediate step - an authentication token we wrote ourselves which has an expiry time which is provided in the URL and used to look up the correct user and device.

**Comment by pcsegal** [comment]
Hi,
I would like to ask: What is the status of this issue? Is there anything planned or any workaround that I could apply as user of the package to eliminate this security concern?
Thank you in advance.

**Comment by moggers87** [comment]
No one has proposed any changes that won't either make things worse or will only work with Django's default auth backend.

IMO, fixing https://github.com/Bouke/django-two-factor-auth/issues/221 would possibly solve this issue too (depending on how it's implemented)

**Comment by ryancausey** [comment]
I've been noodling on this recently and would like some thoughts on the following proposed approach. I don't want to make a PR that ends up being stalled due to not aligning with the architectural desires of this project.

I think if this project switched to a authentication backend approach that this security issue would be solved. I think it would also be able to keep the goal of two-factor auth being pluggable to an extent so that other first factor auth backends can be used. I think the idea would be to have an auth backend per two-factor type since each would ostensibly take different "credentials" to perform the authentication. An ideal login flow would look something like the following:

1. The user hits the login form with credentials for the normal logn backend that is used, be it username and password or some other combination of credentials.
2. The login view, which would have to be the two-factor one or a customized two-factor one that implements the same behavior, would authenticate the user via the normal auth form. This assumes that the auth form being used authenticates the user _but does not log them in_ as the built in auth form does.
3. The view would, if the form is valid, then check if the user requires a second factor of authentication. If they do, the view redirects to the appropriate two-factor auth view. If they do not, the view logs the user in and proceeds as normal.
4. The two-factor auth view would use the appropriate two-factor auth form to validate the second factor credentials. These credentials would match the appropriate backend that would authenticate the user. If the form is valid, I.E. the user authenticated again against the two-factor backend, the user is logged in and things continue as normal from there.

The `user` object cannot be added to the session until the second factor of authentication happens, as the `AbstractBaseUser` sets `is_authenticated` to `True` always, so the mere presence of a user in the session means that it is authenticated. This means that the two-factor backend needs to be able to retrieve the appropriate user based on the passed in credentials or some other metadata. I think this is possible based on the two factor `Device` having a foreign key to user, but I could be wrong.

I was inspired by the [django-axes](https://django-axes.readthedocs.io/en/latest/7_architecture.html) approach where they use an authentication backend to deny a user access if they exceed a certain number of failed login attempts.

**Comment by ryancausey** [comment]
Sorry for the stream of consciousness, but I just though of an additional idea that might make things even more modular.

Rather than having to override the default login views, this could potentially become middleware based wherein a custom two-factor middleware does the "is two-factor required" check after a login request, logs the user back out, and then changes the request to be a redirect to a two-factor view to perform second factor auth and final login. Combined with the auth backend approach, this would potentially become even easier to "bolt on" to existing auth systems.

The trade off is the clunkyness of log out then log back in, but since the log out happens during the same request/response cycle in which the first factor logged the user in, I think it's OK? Definitely need more research to confirm.

Either way I think you could get really crazy and get to the point where `two-factor-auth-core` is just the middleware piece or login view replacements to do the check and redirect. Then the combination of auth backend + view + form for each two-factor method becomes a pluggable and pip-installable `two-factor-auth-*`. I.E. `two-factor-auth-twilio` or `two-factor-auth-OTP`.

**Comment by ryancausey** [comment]
I looked at django-otp a bit. It looks like django-otp deliberately chose not to add an auth backend because that wasn't their project's goal. I can see how that would have influenced the design of the current django-two-factor auth. If this app is really focused on auth, I think it could potentially make sense to add in auth backends for each two-factor provider. It's potentially a bit of maintenance headache if django-otp and its associated plugins move too much, but it might be worth the effort.

**Comment by pbakiewicz** [comment]
How about my proposed solution:

When user successfully authenticates on the `auth` step instead of storing password anywhere, we:
1. Sign user_id using `django.core.signing` module, for even better security use `TimestampSigner` with expiry like 5 mins.
```
from django.core.signing import TimestampSigner
signer = TimestampSigner()
signed_hash = signer.sign(user_id)
```
2. Return signed hash to user as a cookie.
3. On `token` or `backup` step we simply unsign `signed_hash` we got from user's cookie:
```
signer.unsign(value, max_age=300)
```
Doing this we have a certainty that user with given user_id in the last 300s gave us his correct credentials and we can authenticate him if he provides correct token. No passwords stored anywhere.

What do you think?

**Comment by nickcatal** [comment]
@pbakiewicz we have a fix pending in a PR associated with a security alert, tagging this issue and putting it in the README before that fix is live was unintentional.

The going plan now is to verify the user on the 1st page, then set a (changeable) timer on the user within a certain number of minutes they have to finish the entire flow. We're just getting a 500 error in that pull request now and have Github support on it

![Screen Shot 2020-07-07 at 10 56 21 AM](https://user-images.githubusercontent.com/474748/86800867-78c7f900-c041-11ea-8f72-09cf524ae847.png)

**Comment by nickcatal** [comment]
@pbakiewicz we have a fix pending in a PR associated with a security alert, and will likely be filing a CVE, tagging this issue and putting it in the README before that fix, alert, and CVE is live was unintentional.

The going plan now is to verify the user on the 1st page, then set a (changeable) timer on the user within a certain number of minutes they have to finish the entire flow. It's difficult because the package has to support custom backends.

**Comment by pbakiewicz** [comment]
@nickcatal I can't find PR your are referring to. Could you elaborate a bit? 

You mean adding a timer field to auth.User? When user is verified on the 1st page, the timer gets new timestamp, like delta(+5minutes), and in the second step password is not checked at all, only whether the timestamp hasn't expired?

**Comment by nickcatal** [comment]
@pbakiewicz it's a private branch being worked on and will be merged in briefly with a new release

There is a timer set when you submit the `auth` form, it stores a timestamp in your session, and if you do not complete the entire wizard flow within 10 minutes you're sent back to the `auth` form and forced to re-enter your credentials.

**Comment by Bouke** [comment]
The fix is released in 1.12. Thanks to everyone for their contribution on this item and @nickcatal for coming up with the solution that shipped in 1.12!

**Comment by beckedorf** [comment]
Thanks!
I think this issues can now be unpinned.
